### PR TITLE
Feat: Implementation of Elliptic Curve Diffie Hellman Key Exchange.

### PR DESCRIPTION
**Pull request by AshishYUO**
#### Description of Change

Feat: Implementation of [Elliptic Curve Diffie Hellman Key Exchange](https://cryptobook.nakov.com/asymmetric-key-ciphers/ecdh-key-exchange).
Type: Ciphers.
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
There are three files associated with the given topic.

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Great work!
Please fix [`windows-latest`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1479/checks?check_run_id=2314004958) warnings. Mostly, the errors/warnings are displayed at the bottom.

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+ * alicePubKey * bobPrivKey = bobPubKey * alicePrivKey = secret
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+
+#include <algorithm>
+#include <cassert>
+#include <iomanip>
+#include <iostream>
+
+#include "uint256_t.hpp"
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:30]: Add a one-line description of what the library/header is for (see the example below).
```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+}  // namespace elliptic_curve_key_exchange
+}  // namespace ciphers
+
+/**
+ * @brief Function to test the
+ * uint128_t header
+ * @returns void
+ */
+static void uint128_t_tests() {
+    // Tests 1: Operations test
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:175]:
```suggestion
    // 1st test: Operations test
```

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    assert(a * b == uint128_t("282064"));
+    assert(b / a == uint128_t("18"));
+    assert(b % a == uint128_t("116"));
+    assert((a & b) == uint128_t(8));
+    assert((a | b) == 2426);
+    assert((a ^ b) == 2418);
+    assert((a << 64) == uint128_t("2250502776992565297152"));
+    assert((b >> 7) == 18);
+
+    // Tests 2: Operations test
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:188]:
```suggestion
    // 2nd test: Operations test
```

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    assert((a << 64) == uint128_t("227290107637132170748078080907806769152"));
+}
+
+/**
+ * @brief Function to test the
+ * uint256_t header
+ * @returns void
+ */
+static void uint256_t_tests() {
+    // Tests 1: Operations test
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:208]:
```suggestion
    // 1st test: Operations test
```

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    assert(a * b == uint256_t("282064"));
+    assert(b / a == uint256_t("18"));
+    assert(b % a == uint256_t("116"));
+    assert((a & b) == uint256_t(8));
+    assert((a | b) == 2426);
+    assert((a ^ b) == 2418);
+    assert((a << 64) == uint256_t("2250502776992565297152"));
+    assert((b >> 7) == 18);
+
+    // Tests 2: Operations test
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:221]:
```suggestion
    // 2nd test: Operations test
```

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    uint128_t_tests();  // running predefined 128-bit unsigned integer tests.
+    uint256_t_tests();  // running predefined 256-bit unsigned integer tests.
+    test();             // running predefined tests
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:300]:
```suggestion
    test();             // running self-test implementations
```

File: ciphers/elliptic_curve_key_exchange.cpp [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    assert(alice_shared_key == bob_shared_key);
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    uint128_t_tests();  // running predefined 128-bit unsigned integer tests.
+    uint256_t_tests();  // running predefined 256-bit unsigned integer tests.
```
Comment (Panquesito7) [inline ciphers/elliptic_curve_key_exchange.cpp:299]:
```suggestion
    uint128_t_tests();  // running predefined 128-bit unsigned integer tests
    uint256_t_tests();  // running predefined 256-bit unsigned integer tests
```

File: ciphers/uint128_t.hpp [hunk 8]
```diff
@@ -0,0 +1,1075 @@
+/**
+ * @file uint128_t.hpp
```
Comment (Panquesito7) [inline ciphers/uint128_t.hpp:2]:
```suggestion
 * @file
```

File: ciphers/uint128_t.hpp [hunk 9]
```diff
@@ -0,0 +1,1075 @@
...
+ * @details Implementation of 128-bit unsigned integers.
+ * @note The implementation can be flagged as not completed. This header is used
+ * with enough operations as a part of bigger integer types 256-bit integer.
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+
+#include <algorithm>
+#include <ostream>
+#include <string>
+#include <utility>
```
Comment (Panquesito7) [inline ciphers/uint128_t.hpp:13]: Add a one-line description of what the library/header is for (see the example below).
```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: ciphers/uint128_t.hpp [hunk 10]
```diff
@@ -0,0 +1,1075 @@
...
+    }
+    if (carry) {
+        third.push_back('1');
+    }
+    std::reverse(third.begin(), third.end());
+    return third;
+}
+/**
+ * @class uint128_t
+ * @details 128-bit numbers.
```
Comment (Panquesito7) [inline ciphers/uint128_t.hpp:57]:
```suggestion
 * @brief 128-bit numbers.
```

File: ciphers/uint256_t.hpp [hunk 11]
```diff
@@ -0,0 +1,847 @@
+/**
+ * @file uint256_t.hpp
```
Comment (Panquesito7) [inline ciphers/uint256_t.hpp:2]:
```suggestion
 * @file
```

File: ciphers/uint256_t.hpp [hunk 12]
```diff
@@ -0,0 +1,847 @@
...
+ * @details Implementation of 256-bit unsigned integers.
+ * @note The implementation can be flagged as not completed. This header is used
+ * with enough operations to demonstrate the usage of ECDH (Elliptic Curve
+ * Diffie-Hellman) Key exchange.
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+#include <string>
+#include <utility>
+
+#include "uint128_t.hpp"
```
Comment (Panquesito7) [inline ciphers/uint256_t.hpp:13]: Add a one-line description of what the library/header is for (see the example below).
```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: ciphers/uint256_t.hpp [hunk 13]
```diff
@@ -0,0 +1,847 @@
...
+
+template <>
+struct std::is_arithmetic<uint256_t> : std::true_type {};
+
+template <>
+struct std::is_unsigned<uint256_t> : std::true_type {};
+
+/**
+ * @class uint256_t
+ * @details 256-bit number class.
```
Comment (Panquesito7) [inline ciphers/uint256_t.hpp:31]:
```suggestion
 * @brief 256-bit number class
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: ciphers/uint256_t.hpp [hunk 14]
```diff
@@ -0,0 +1,1036 @@
...
+        }
+    }
+
+ public:
+    // Constructors
+    uint256_t() = default;
+
+    template <typename T, typename = typename std::enable_if<
+                              std::is_integral<T>::value, T>::type>
+    explicit uint256_t(T low) : s(low), f(0) {}
```
Comment (Panquesito7) [inline ciphers/uint256_t.hpp:70]: Add a one-line description of what these variables do or are for.

**Review by Panquesito7 — APPROVED** [review]
Awesome work! I like how you've improved and made the code so well. 😄
Code and documentation are pretty well-made. Thank you for your contribution! 👍 🎉

**Review by Panquesito7 — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]

**Pull request merged**